Plans: Use `/checkout/:site/:plan` as a route to checkout with a plan

drewblaisdell opened the pull request:
Fixes #2021.

We need a route that we can link to that takes the user to the checkout process for a specific plan from outside of Calypso. Previously, we used `PlansSelect` (`/plans/select/:site/:plan`), but:

- This component doesn't have a loading state.
- We're linking to the checkout process, so the route should live under `/checkout` instead of `/plans/select`.

This PR:

- Adds an optional path fragment to the checkout route: `/checkout/:site/:plan?`
- Updates the checkout component to add a plan to the cart if the `:plan?` fragment is present.
- Removes `PlanSelect`.
- Redirects `/plans/select/:site/:plan` to `/checkout/:site/:plan`.

**Testing**
- Visit `/checkout/:site/premium` or `/checkout/:site/business` for a site without premium/business or with the premium/business trial.
- Assert that you see a loading screen, that the correct plan is then added to the cart, and that you can complete the checkout.

- [x] Product review
- [x] Code review

Tug commented on client/my-sites/upgrades/checkout/checkout.jsx:
```diff
@@ -34,6 +34,10 @@ module.exports = React.createClass( {
 			this.trackPageView();
+
+			if ( this.props.plan ) {
+				this.addPlanToCart()
```
Could use a semicolon at the end of the line (same for L54)

Tug commented on client/my-sites/upgrades/controller.jsx:
```diff
@@ -178,6 +179,8 @@ module.exports = {
 				<CheckoutData>
 					<Checkout
 						cards={ storedCards }
+						plan={ context.params.plan }
```
Maybe use `planName` or `selectedPlanName` instead of `plan` to show implicitly that it holds a string with the short name for the plan.

Tug commented:
I left a couple of comments, otherwise I think it's ready for merge :ship:

Tug commented:
And product tested :+1:

drewblaisdell commented:
@Tug Thanks for looking at this. I updated the initial commit based on your notes, and added a commit to indicate that `/plans/select/:site/:plan` is deprecated in the plans README.

Tug commented:
:+1:
